projector/binding.go

```diff
@@ -42,69 +52,354 @@ func New(mappingSource MappingSource) ServiceBindingProjector {
 	// rather than attempt to merge an existing binding, unproject it
-	if err := i.unproject(ctx, binding, mpt); err != nil {
-		return err
+	p.unproject(binding, mpt)
+
+	if p.secretName(binding) == "" {
+		// no secret to bind
+		return
 	}
+	p.projectVolume(binding, mpt)
+	for i := range mpt.Containers {
+		p.projectContainer(binding, mpt, &mpt.Containers[i])
+	}
+}
 
+func (p *serviceBindingProjector) unproject(binding *servicebindingv1alpha3.ServiceBinding, mpt *metaPodTemplate) {
+	p.unprojectVolume(binding, mpt)
 	for i := range mpt.Containers {
-		c := &mpt.Containers[i]
-		// TODO skip container if not allowed
-
-		serviceBindingRoot := ""
-		for _, e := range c.Env {
-			if e.Name == ServiceBindingRootEnv {
-				serviceBindingRoot = e.Value
-				break
-			}
+		p.unprojectContainer(binding, mpt, &mpt.Containers[i])
+	}
+
+	// cleanup annotations
+	delete(mpt.Annotations, p.secretAnnotationName(binding))
+	delete(mpt.Annotations, p.typeAnnotationName(binding))
+	delete(mpt.Annotations, p.providerAnnotationName(binding))
+}
+
+func (p *serviceBindingProjector) projectVolume(binding *servicebindingv1alpha3.ServiceBinding, mpt *metaPodTemplate) {
+	volume := corev1.Volume{
+		Name: p.volumeName(binding),
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				Sources: []corev1.VolumeProjection{
+					{
+						Secret: &corev1.SecretProjection{
+							LocalObjectReference: corev1.LocalObjectReference{
+								Name: p.secretAnnotation(binding, mpt),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	if binding.Spec.Type != "" {
+		volume.VolumeSource.Projected.Sources = append(volume.VolumeSource.Projected.Sources,
+			corev1.VolumeProjection{
+				DownwardAPI: &corev1.DownwardAPIProjection{
+					Items: []corev1.DownwardAPIVolumeFile{
+						{
+							Path: "type",
+							FieldRef: &corev1.ObjectFieldSelector{
+								FieldPath: fmt.Sprintf("metadata.annotations['%s']", p.typeAnnotation(binding, mpt)),
+							},
+						},
+					},
+				},
+			},
+		)
+	}
+	if binding.Spec.Provider != "" {
+		volume.VolumeSource.Projected.Sources = append(volume.VolumeSource.Projected.Sources,
+			corev1.VolumeProjection{
+				DownwardAPI: &corev1.DownwardAPIProjection{
+					Items: []corev1.DownwardAPIVolumeFile{
+						{
+							Path: "provider",
+							FieldRef: &corev1.ObjectFieldSelector{
+								FieldPath: fmt.Sprintf("metadata.annotations['%s']", p.providerAnnotation(binding, mpt)),
+							},
+						},
+					},
+				},
+			},
+		)
+	}
+
+	mpt.Volumes = append(mpt.Volumes, volume)
+
+	// sort all injected volumes
+	sort.SliceStable(mpt.Volumes, func(i, j int) bool {
+		in := mpt.Volumes[i].Name
+		jn := mpt.Volumes[j].Name
+		if !strings.HasPrefix(in, VolumePrefix) || !strings.HasPrefix(jn, VolumePrefix) {
+			return false
+		}
+		return in < jn
+	})
+}
+
+func (p *serviceBindingProjector) unprojectVolume(binding *servicebindingv1alpha3.ServiceBinding, mpt *metaPodTemplate) {
```

Comment:
This is the scenario I have in mind.. please apply the following diff:

```diff
diff --git a/projector/binding_test.go b/projector/binding_test.go
index 4d7c803..ea8c04d 100644
--- a/projector/binding_test.go
+++ b/projector/binding_test.go
@@ -1498,6 +1498,12 @@ func TestBinding(t *testing.T) {
 										},
 									},
 								},
+								{
+									Name: "preexisting2",
+									VolumeSource: corev1.VolumeSource{
+										EmptyDir: &corev1.EmptyDirVolumeSource{},
+									},
+								},
 							},
 							Containers: []corev1.Container{
 								{
@@ -1630,14 +1636,14 @@ func TestBinding(t *testing.T) {
 									},
 								},
 								{
-									Name: "servicebinding-26894874-4719-4802-8f43-8ceed127b4c2",
+									Name: "servicebinding-33333333-3333-3333-3333-333333333333",
 									VolumeSource: corev1.VolumeSource{
 										Projected: &corev1.ProjectedVolumeSource{
 											Sources: []corev1.VolumeProjection{
 												{
 													Secret: &corev1.SecretProjection{
 														LocalObjectReference: corev1.LocalObjectReference{
-															Name: secretName,
+															Name: "secret-3",
 														},
 													},
 												},
@@ -1646,14 +1652,20 @@ func TestBinding(t *testing.T) {
 									},
 								},
 								{
-									Name: "servicebinding-33333333-3333-3333-3333-333333333333",
+									Name: "preexisting2",
+									VolumeSource: corev1.VolumeSource{
+										EmptyDir: &corev1.EmptyDirVolumeSource{},
+									},
+								},
+								{
+									Name: "servicebinding-26894874-4719-4802-8f43-8ceed127b4c2",
 									VolumeSource: corev1.VolumeSource{
 										Projected: &corev1.ProjectedVolumeSource{
 											Sources: []corev1.VolumeProjection{
 												{
 													Secret: &corev1.SecretProjection{
 														LocalObjectReference: corev1.LocalObjectReference{
-															Name: "secret-3",
+															Name: secretName,
 														},
 													},
 												},

```

Volumes will not be sorted properly, i.e. the new volume will be added after the one called `preexisting2`. Should it be like that?

I could agree that `Project` is idempotent, but it would be nice that applying a service binding twice does not change the workload at all, even not changing the order of volumes, etc. However, this does happen, please apply the following diff for demonstration:

```diff
diff --git a/projector/binding_test.go b/projector/binding_test.go
index 4d7c803..48a7e09 100644
--- a/projector/binding_test.go
+++ b/projector/binding_test.go
@@ -1735,6 +1735,210 @@ func TestBinding(t *testing.T) {
 				},
 			},
 		},
+		{
+			name:    "apply binding should be idempotent",
+			mapping: NewStaticMapping(&servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate{}),
+			binding: &servicebindingv1alpha3.ServiceBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					UID: uid,
+				},
+				Status: servicebindingv1alpha3.ServiceBindingStatus{
+					Binding: &servicebindingv1alpha3.ServiceBindingSecretReference{
+						Name: secretName,
+					},
+				},
+			},
+			workload: &appsv1.Deployment{
+				Spec: appsv1.DeploymentSpec{
+					Template: corev1.PodTemplateSpec{
+						ObjectMeta: metav1.ObjectMeta{
+							Annotations: map[string]string{
+								"projector.servicebinding.io/secret-11111111-1111-1111-1111-111111111111": "secret-1",
+								"projector.servicebinding.io/secret-26894874-4719-4802-8f43-8ceed127b4c2": "my-secret",
+							},
+						},
+						Spec: corev1.PodSpec{
+							Volumes: []corev1.Volume{
+								{
+									Name: "preexisting",
+									VolumeSource: corev1.VolumeSource{
+										EmptyDir: &corev1.EmptyDirVolumeSource{},
+									},
+								},
+								{
+									Name: "servicebinding-11111111-1111-1111-1111-111111111111",
+									VolumeSource: corev1.VolumeSource{
+										Projected: &corev1.ProjectedVolumeSource{
+											Sources: []corev1.VolumeProjection{
+												{
+													Secret: &corev1.SecretProjection{
+														LocalObjectReference: corev1.LocalObjectReference{
+															Name: "secret-1",
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+								{
+									Name: "servicebinding-26894874-4719-4802-8f43-8ceed127b4c2",
+									VolumeSource: corev1.VolumeSource{
+										Projected: &corev1.ProjectedVolumeSource{
+											Sources: []corev1.VolumeProjection{
+												{
+													Secret: &corev1.SecretProjection{
+														LocalObjectReference: corev1.LocalObjectReference{
+															Name: "my-secret",
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+								{
+									Name: "preexisting2",
+									VolumeSource: corev1.VolumeSource{
+										EmptyDir: &corev1.EmptyDirVolumeSource{},
+									},
+								},
+							},
+							Containers: []corev1.Container{
+								{
+									Env: []corev1.EnvVar{
+										{
+											Name:  "PREEXISTING",
+											Value: "env",
+										},
+										{
+											Name:  "SERVICE_BINDING_ROOT",
+											Value: "/bindings",
+										},
+									},
+									VolumeMounts: []corev1.VolumeMount{
+										{
+											Name:      "preexisting",
+											MountPath: "/var/mount",
+										},
+										{
+											Name:      "servicebinding-11111111-1111-1111-1111-111111111111",
+											ReadOnly:  true,
+											MountPath: "/bindings/binding-1",
+										},
+										{
+											Name:      "servicebinding-26894874-4719-4802-8f43-8ceed127b4c2",
+											ReadOnly:  true,
+											MountPath: "/bindings/my-binding",
+										},
+										{
+											Name:      "preexisting2",
+											MountPath: "/var/mount",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: &appsv1.Deployment{
+				Spec: appsv1.DeploymentSpec{
+					Template: corev1.PodTemplateSpec{
+						ObjectMeta: metav1.ObjectMeta{
+							Annotations: map[string]string{
+								"projector.servicebinding.io/secret-11111111-1111-1111-1111-111111111111": "secret-1",
+								"projector.servicebinding.io/secret-26894874-4719-4802-8f43-8ceed127b4c2": "my-secret",
+							},
+						},
+						Spec: corev1.PodSpec{
+							Volumes: []corev1.Volume{
+								{
+									Name: "preexisting",
+									VolumeSource: corev1.VolumeSource{
+										EmptyDir: &corev1.EmptyDirVolumeSource{},
+									},
+								},
+								{
+									Name: "servicebinding-11111111-1111-1111-1111-111111111111",
+									VolumeSource: corev1.VolumeSource{
+										Projected: &corev1.ProjectedVolumeSource{
+											Sources: []corev1.VolumeProjection{
+												{
+													Secret: &corev1.SecretProjection{
+														LocalObjectReference: corev1.LocalObjectReference{
+															Name: "secret-1",
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+
+								{
+									Name: "preexisting2",
+									VolumeSource: corev1.VolumeSource{
+										EmptyDir: &corev1.EmptyDirVolumeSource{},
+									},
+								},
+								{
+									Name: "servicebinding-26894874-4719-4802-8f43-8ceed127b4c2",
+									VolumeSource: corev1.VolumeSource{
+										Projected: &corev1.ProjectedVolumeSource{
+											Sources: []corev1.VolumeProjection{
+												{
+													Secret: &corev1.SecretProjection{
+														LocalObjectReference: corev1.LocalObjectReference{
+															Name: "my-secret",
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+							Containers: []corev1.Container{
+								{
+									Env: []corev1.EnvVar{
+										{
+											Name:  "PREEXISTING",
+											Value: "env",
+										},
+										{
+											Name:  "SERVICE_BINDING_ROOT",
+											Value: "/bindings",
+										},
+									},
+									VolumeMounts: []corev1.VolumeMount{
+										{
+											Name:      "preexisting",
+											MountPath: "/var/mount",
+										},
+										{
+											Name:      "servicebinding-11111111-1111-1111-1111-111111111111",
+											ReadOnly:  true,
+											MountPath: "/bindings/binding-1",
+										},
+
+										{
+											Name:      "preexisting2",
+											MountPath: "/var/mount",
+										},
+										{
+											Name:      "servicebinding-26894874-4719-4802-8f43-8ceed127b4c2",
+											ReadOnly:  true,
+											MountPath: "/bindings",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
 		{
 			name: "invalid container jsonpath",
 			mapping: NewStaticMapping(&servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate{

```

I believe that changing order is going to restart workload pods, unnecessary.

Reply:
> it would be nice that applying a service binding twice does not change the workload at all, even not changing the order of volumes

It already is, the ordering is just different than what you're proposing. Only the volumes/volumeMounts/env injected by this projector will have their order changed.